engine/access/state_stream/backend/handler.go

```diff
@@ -165,6 +170,24 @@ func (h *Handler) SubscribeEvents(request *executiondata.SubscribeEventsRequest,
 			return status.Errorf(codes.Internal, "unexpected response type: %T", v)
 		}
 
+		// check if there are any events in the response. if not, do not send a message unless the last
+		// response was more than HeartbeatInterval blocks ago
+		if len(resp.Events) == 0 {
+			blocksSinceLastMessage++
+
+			heartbeatInterval := uint64(0)
+			if request.HeartbeatInterval == 0 {
```

Comment:
this logic can be greatly simplified if heartbeat init is moved in the beginning of the function